// @samchon/openapi/lib/composers/llm/ChatGptSchemaComposer.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatGptSchemaComposer = void 0;
var ChatGptTypeChecker_1 = require("../../utils/ChatGptTypeChecker");
var LlmTypeCheckerV3_1_1 = require("../../utils/LlmTypeCheckerV3_1");
var NamingConvention_1 = require("../../utils/NamingConvention");
var OpenApiTypeChecker_1 = require("../../utils/OpenApiTypeChecker");
var LlmDescriptionInverter_1 = require("./LlmDescriptionInverter");
var LlmSchemaV3_1Composer_1 = require("./LlmSchemaV3_1Composer");
var ChatGptSchemaComposer;
(function (ChatGptSchemaComposer) {
    /**
     * @internal
     */
    ChatGptSchemaComposer.IS_DEFS = true;
    /* -----------------------------------------------------------
      CONVERTERS
    ----------------------------------------------------------- */
    ChatGptSchemaComposer.parameters = function (props) {
        var e_1, _a;
        var _b;
        var _c;
        // polyfill
        (_b = (_c = props.config).strict) !== null && _b !== void 0 ? _b : (_c.strict = false);
        // validate
        var result = LlmSchemaV3_1Composer_1.LlmSchemaV3_1Composer.parameters(__assign(__assign({}, props), { config: {
                reference: props.config.reference,
                constraint: false,
            }, validate: props.config.strict === true ? validateStrict : undefined }));
        if (result.success === false)
            return result;
        try {
            // returns with transformation
            for (var _d = __values(Object.keys(result.value.$defs)), _e = _d.next(); !_e.done; _e = _d.next()) {
                var key = _e.value;
                result.value.$defs[key] = transform({
                    config: props.config,
                    schema: result.value.$defs[key],
                });
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return {
            success: true,
            value: transform({
                config: props.config,
                schema: result.value,
            }),
        };
    };
    ChatGptSchemaComposer.schema = function (props) {
        var e_2, _a;
        var _b;
        var _c;
        // polyfill
        (_b = (_c = props.config).strict) !== null && _b !== void 0 ? _b : (_c.strict = false);
        // validate
        var oldbie = new Set(Object.keys(props.$defs));
        var result = LlmSchemaV3_1Composer_1.LlmSchemaV3_1Composer.schema(__assign(__assign({}, props), { config: {
                reference: props.config.reference,
                constraint: false,
            }, validate: props.config.strict === true ? validateStrict : undefined }));
        if (result.success === false)
            return result;
        try {
            // returns with transformation
            for (var _d = __values(Object.keys(props.$defs)), _e = _d.next(); !_e.done; _e = _d.next()) {
                var key = _e.value;
                if (oldbie.has(key) === false)
                    props.$defs[key] = transform({
                        config: props.config,
                        schema: props.$defs[key],
                    });
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return {
            success: true,
            value: transform({
                config: props.config,
                schema: result.value,
            }),
        };
    };
    var validateStrict = function (schema, accessor) {
        var e_3, _a;
        var _b, _c;
        var reasons = [];
        if (OpenApiTypeChecker_1.OpenApiTypeChecker.isObject(schema)) {
            if (!!schema.additionalProperties)
                reasons.push({
                    schema: schema,
                    accessor: "".concat(accessor, ".additionalProperties"),
                    message: "ChatGPT does not allow additionalProperties in strict mode, the dynamic key typed object.",
                });
            try {
                for (var _d = __values(Object.keys((_b = schema.properties) !== null && _b !== void 0 ? _b : {})), _e = _d.next(); !_e.done; _e = _d.next()) {
                    var key = _e.value;
                    if (((_c = schema.required) === null || _c === void 0 ? void 0 : _c.includes(key)) === false)
                        reasons.push({
                            schema: schema,
                            accessor: "".concat(accessor, ".properties.").concat(key),
                            message: "ChatGPT does not allow optional properties in strict mode.",
                        });
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        return reasons;
    };
    var transform = function (props) {
        var _a;
        var union = [];
        var attribute = __assign({ title: props.schema.title, description: props.schema.description, example: props.schema.example, examples: props.schema.examples }, Object.fromEntries(Object.entries(ChatGptSchemaComposer.schema).filter(function (_a) {
            var _b = __read(_a, 2), key = _b[0], value = _b[1];
            return key.startsWith("x-") && value !== undefined;
        })));
        var visit = function (input) {
            if (LlmTypeCheckerV3_1_1.LlmTypeCheckerV3_1.isOneOf(input))
                input.oneOf.forEach(visit);
            else if (LlmTypeCheckerV3_1_1.LlmTypeCheckerV3_1.isArray(input))
                union.push(__assign(__assign({}, input), { items: transform({
                        config: props.config,
                        schema: input.items,
                    }) }));
            else if (LlmTypeCheckerV3_1_1.LlmTypeCheckerV3_1.isObject(input))
                union.push(__assign(__assign({}, input), { properties: Object.fromEntries(Object.entries(input.properties).map(function (_a) {
                        var _b = __read(_a, 2), key = _b[0], value = _b[1];
                        return [
                            key,
                            transform({
                                config: props.config,
                                schema: value,
                            }),
                        ];
                    })), additionalProperties: props.config.strict === true
                        ? false
                        : typeof input.additionalProperties === "object" &&
                            input.additionalProperties !== null
                            ? transform({
                                config: props.config,
                                schema: input.additionalProperties,
                            })
                            : input.additionalProperties }));
            else if (LlmTypeCheckerV3_1_1.LlmTypeCheckerV3_1.isConstant(input) === false)
                union.push(input);
        };
        var visitConstant = function (input) {
            var insert = function (value) {
                var _a;
                var matched = union.find(function (u) {
                    return (u === null || u === void 0 ? void 0 : u.type) ===
                        typeof value;
                });
                if (matched !== undefined) {
                    (_a = matched.enum) !== null && _a !== void 0 ? _a : (matched.enum = []);
                    matched.enum.push(value);
                }
                else
                    union.push({
                        type: typeof value,
                        enum: [value],
                    });
            };
            if (OpenApiTypeChecker_1.OpenApiTypeChecker.isConstant(input))
                insert(input.const);
            else if (OpenApiTypeChecker_1.OpenApiTypeChecker.isOneOf(input))
                input.oneOf.forEach(visitConstant);
        };
        visit(props.schema);
        visitConstant(props.schema);
        if (union.length === 0)
            return __assign(__assign({}, attribute), { type: undefined });
        else if (union.length === 1)
            return __assign(__assign(__assign({}, attribute), union[0]), { description: ChatGptTypeChecker_1.ChatGptTypeChecker.isReference(union[0])
                    ? undefined
                    : ((_a = union[0].description) !== null && _a !== void 0 ? _a : attribute.description) });
        return __assign(__assign({}, attribute), { anyOf: union.map(function (u) { return (__assign(__assign({}, u), { description: ChatGptTypeChecker_1.ChatGptTypeChecker.isReference(u)
                    ? undefined
                    : u.description })); }) });
    };
    /* -----------------------------------------------------------
      SEPARATORS
    ----------------------------------------------------------- */
    ChatGptSchemaComposer.separateParameters = function (props) {
        var e_4, _a;
        var _b, _c;
        var convention = (_b = props.convention) !== null && _b !== void 0 ? _b : (function (key, type) { return "".concat(key, ".").concat(NamingConvention_1.NamingConvention.capitalize(type)); });
        var _d = __read(separateObject({
            predicate: props.predicate,
            convention: convention,
            $defs: props.parameters.$defs,
            schema: props.parameters,
        }), 2), llm = _d[0], human = _d[1];
        if (llm === null || human === null)
            return {
                llm: (_c = llm) !== null && _c !== void 0 ? _c : {
                    type: "object",
                    properties: {},
                    required: [],
                    additionalProperties: false,
                    $defs: {},
                },
                human: human,
            };
        var output = {
            llm: __assign(__assign({}, llm), { $defs: Object.fromEntries(Object.entries(props.parameters.$defs).filter(function (_a) {
                    var _b = __read(_a, 1), key = _b[0];
                    return key.endsWith(".Llm");
                })), additionalProperties: false }),
            human: __assign(__assign({}, human), { $defs: Object.fromEntries(Object.entries(props.parameters.$defs).filter(function (_a) {
                    var _b = __read(_a, 1), key = _b[0];
                    return key.endsWith(".Human");
                })), additionalProperties: false }),
        };
        try {
            for (var _e = __values(Object.keys(props.parameters.$defs)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var key = _f.value;
                if (key.endsWith(".Llm") === false && key.endsWith(".Human") === false)
                    delete props.parameters.$defs[key];
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_a = _e.return)) _a.call(_e);
            }
            finally { if (e_4) throw e_4.error; }
        }
        return output;
    };
    var separateStation = function (props) {
        if (props.predicate(props.schema) === true)
            return [null, props.schema];
        else if (ChatGptTypeChecker_1.ChatGptTypeChecker.isUnknown(props.schema) ||
            ChatGptTypeChecker_1.ChatGptTypeChecker.isAnyOf(props.schema))
            return [props.schema, null];
        else if (ChatGptTypeChecker_1.ChatGptTypeChecker.isObject(props.schema))
            return separateObject({
                predicate: props.predicate,
                convention: props.convention,
                $defs: props.$defs,
                schema: props.schema,
            });
        else if (ChatGptTypeChecker_1.ChatGptTypeChecker.isArray(props.schema))
            return separateArray({
                predicate: props.predicate,
                convention: props.convention,
                $defs: props.$defs,
                schema: props.schema,
            });
        else if (ChatGptTypeChecker_1.ChatGptTypeChecker.isReference(props.schema))
            return separateReference({
                predicate: props.predicate,
                convention: props.convention,
                $defs: props.$defs,
                schema: props.schema,
            });
        return [props.schema, null];
    };
    var separateArray = function (props) {
        var _a = __read(separateStation({
            predicate: props.predicate,
            convention: props.convention,
            $defs: props.$defs,
            schema: props.schema.items,
        }), 2), x = _a[0], y = _a[1];
        return [
            x !== null
                ? __assign(__assign({}, props.schema), { items: x }) : null,
            y !== null
                ? __assign(__assign({}, props.schema), { items: y }) : null,
        ];
    };
    var separateObject = function (props) {
        var e_5, _a;
        var _b, _c;
        // EMPTY OBJECT
        if (Object.keys((_b = props.schema.properties) !== null && _b !== void 0 ? _b : {}).length === 0 &&
            !!props.schema.additionalProperties === false)
            return [props.schema, null];
        var llm = __assign(__assign({}, props.schema), { properties: {}, additionalProperties: props.schema.additionalProperties });
        var human = __assign(__assign({}, props.schema), { properties: {} });
        try {
            for (var _d = __values(Object.entries((_c = props.schema.properties) !== null && _c !== void 0 ? _c : {})), _e = _d.next(); !_e.done; _e = _d.next()) {
                var _f = __read(_e.value, 2), key = _f[0], value = _f[1];
                var _g = __read(separateStation({
                    predicate: props.predicate,
                    convention: props.convention,
                    $defs: props.$defs,
                    schema: value,
                }), 2), x = _g[0], y = _g[1];
                if (x !== null)
                    llm.properties[key] = x;
                if (y !== null)
                    human.properties[key] = y;
            }
        }
        catch (e_5_1) { e_5 = { error: e_5_1 }; }
        finally {
            try {
                if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
            }
            finally { if (e_5) throw e_5.error; }
        }
        if (typeof props.schema.additionalProperties === "object" &&
            props.schema.additionalProperties !== null) {
            var _h = __read(separateStation({
                predicate: props.predicate,
                convention: props.convention,
                $defs: props.$defs,
                schema: props.schema.additionalProperties,
            }), 2), dx = _h[0], dy = _h[1];
            llm.additionalProperties = dx !== null && dx !== void 0 ? dx : false;
            human.additionalProperties = dy !== null && dy !== void 0 ? dy : false;
        }
        return [
            !!Object.keys(llm.properties).length || !!llm.additionalProperties
                ? shrinkRequired(llm)
                : null,
            !!Object.keys(human.properties).length || human.additionalProperties
                ? shrinkRequired(human)
                : null,
        ];
    };
    var separateReference = function (props) {
        var _a, _b, _c, _d, _e;
        var key = props.schema.$ref.split("#/$defs/")[1];
        var humanKey = props.convention(key, "human");
        var llmKey = props.convention(key, "llm");
        // FIND EXISTING
        if (((_a = props.$defs) === null || _a === void 0 ? void 0 : _a[humanKey]) || ((_b = props.$defs) === null || _b === void 0 ? void 0 : _b[llmKey]))
            return [
                ((_c = props.$defs) === null || _c === void 0 ? void 0 : _c[llmKey])
                    ? __assign(__assign({}, props.schema), { $ref: "#/$defs/".concat(llmKey) }) : null,
                ((_d = props.$defs) === null || _d === void 0 ? void 0 : _d[humanKey])
                    ? __assign(__assign({}, props.schema), { $ref: "#/$defs/".concat(humanKey) }) : null,
            ];
        // PRE-ASSIGNMENT
        props.$defs[llmKey] = {};
        props.$defs[humanKey] = {};
        // DO COMPOSE
        var schema = (_e = props.$defs) === null || _e === void 0 ? void 0 : _e[key];
        var _f = __read(separateStation({
            predicate: props.predicate,
            convention: props.convention,
            $defs: props.$defs,
            schema: schema,
        }), 2), llm = _f[0], human = _f[1];
        // ONLY ONE
        if (llm === null || human === null) {
            delete props.$defs[llmKey];
            delete props.$defs[humanKey];
            return llm === null ? [null, props.schema] : [props.schema, null];
        }
        // BOTH OF THEM
        return [
            llm !== null
                ? __assign(__assign({}, props.schema), { $ref: "#/$defs/".concat(llmKey) }) : null,
            human !== null
                ? __assign(__assign({}, props.schema), { $ref: "#/$defs/".concat(humanKey) }) : null,
        ];
    };
    var shrinkRequired = function (s) {
        s.required = s.required.filter(function (key) { var _a; return ((_a = s.properties) === null || _a === void 0 ? void 0 : _a[key]) !== undefined; });
        return s;
    };
    /* -----------------------------------------------------------
      INVERTERS
    ----------------------------------------------------------- */
    ChatGptSchemaComposer.invert = function (props) {
        var union = [];
        var attribute = __assign(__assign({ title: props.schema.title, description: props.schema.description }, Object.fromEntries(Object.entries(props.schema).filter(function (_a) {
            var _b = __read(_a, 2), key = _b[0], value = _b[1];
            return key.startsWith("x-") && value !== undefined;
        }))), { example: props.schema.example, examples: props.schema.examples });
        var next = function (schema) {
            return ChatGptSchemaComposer.invert({
                components: props.components,
                $defs: props.$defs,
                schema: schema,
            });
        };
        var visit = function (schema) {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j;
            var _k;
            if (ChatGptTypeChecker_1.ChatGptTypeChecker.isArray(schema))
                union.push(__assign(__assign(__assign({}, LlmDescriptionInverter_1.LlmDescriptionInverter.array((_a = schema.description) !== null && _a !== void 0 ? _a : "")), schema), { items: next(schema.items) }));
            else if (ChatGptTypeChecker_1.ChatGptTypeChecker.isObject(schema))
                union.push(__assign(__assign({}, schema), { properties: Object.fromEntries(Object.entries(schema.properties).map(function (_a) {
                        var _b = __read(_a, 2), key = _b[0], value = _b[1];
                        return [
                            key,
                            next(value),
                        ];
                    })), additionalProperties: typeof schema.additionalProperties === "object" &&
                        schema.additionalProperties !== null
                        ? next(schema.additionalProperties)
                        : schema.additionalProperties }));
            else if (ChatGptTypeChecker_1.ChatGptTypeChecker.isAnyOf(schema))
                schema.anyOf.forEach(visit);
            else if (ChatGptTypeChecker_1.ChatGptTypeChecker.isReference(schema)) {
                var key = schema.$ref.split("#/$defs/")[1];
                if (((_b = props.components.schemas) === null || _b === void 0 ? void 0 : _b[key]) === undefined) {
                    (_c = (_k = props.components).schemas) !== null && _c !== void 0 ? _c : (_k.schemas = {});
                    props.components.schemas[key] = {};
                    props.components.schemas[key] = next((_d = props.$defs[key]) !== null && _d !== void 0 ? _d : {});
                }
                union.push(__assign(__assign({}, schema), { $ref: "#/components/schemas/".concat(key) }));
            }
            else if (ChatGptTypeChecker_1.ChatGptTypeChecker.isBoolean(schema))
                if (!!((_e = schema.enum) === null || _e === void 0 ? void 0 : _e.length))
                    schema.enum.forEach(function (v) {
                        return union.push({
                            const: v,
                        });
                    });
                else
                    union.push(schema);
            else if (ChatGptTypeChecker_1.ChatGptTypeChecker.isInteger(schema) ||
                ChatGptTypeChecker_1.ChatGptTypeChecker.isNumber(schema))
                if (!!((_f = schema.enum) === null || _f === void 0 ? void 0 : _f.length))
                    schema.enum.forEach(function (v) {
                        return union.push({
                            const: v,
                        });
                    });
                else
                    union.push(__assign(__assign(__assign({}, LlmDescriptionInverter_1.LlmDescriptionInverter.numeric((_g = schema.description) !== null && _g !== void 0 ? _g : "")), schema), { enum: undefined }));
            else if (ChatGptTypeChecker_1.ChatGptTypeChecker.isString(schema))
                if (!!((_h = schema.enum) === null || _h === void 0 ? void 0 : _h.length))
                    schema.enum.forEach(function (v) {
                        return union.push({
                            const: v,
                        });
                    });
                else
                    union.push(__assign(__assign(__assign({}, LlmDescriptionInverter_1.LlmDescriptionInverter.string((_j = schema.description) !== null && _j !== void 0 ? _j : "")), schema), { enum: undefined }));
            else
                union.push(__assign({}, schema));
        };
        visit(props.schema);
        return __assign(__assign({}, (union.length === 0
            ? { type: undefined }
            : union.length === 1
                ? __assign({}, union[0]) : { oneOf: union.map(function (u) { return (__assign(__assign({}, u), { nullable: undefined })); }) })), attribute);
    };
})(ChatGptSchemaComposer || (exports.ChatGptSchemaComposer = ChatGptSchemaComposer = {}));
