// @lynx-js/template-webpack-plugin/lib/css/encode.js
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
import { removeFunctionWhiteSpace } from '@lynx-js/css-serializer/dist/plugins/removeFunctionWhiteSpace.js';
import { cssChunksToMap } from './cssChunksToMap.js';
export async function encodeCSS(cssChunks, { enableCSSSelector, enableRemoveCSSScope, targetSdkVersion, }, plugins = [removeFunctionWhiteSpace()], 
// biome-ignore lint/suspicious/noExplicitAny: <explanation>
encode = (options) => {
    const buffer = Buffer.from(JSON.stringify(options));
    return Promise.resolve({
        buffer,
    });
}) {
    const css = cssChunksToMap(cssChunks, plugins);
    const encodeOptions = {
        compilerOptions: {
            // Do not remove this, it will crash :)
            enableFiberArch: true,
            useLepusNG: true,
            bundleModuleMode: 'ReturnByFunction',
            enableCSSSelector,
            targetSdkVersion,
        },
        sourceContent: {
            appType: 'card',
            config: {
                lepusStrict: true,
                enableRemoveCSSScope,
            },
        },
        css,
    };
    const { buffer } = await encode(encodeOptions);
    return buffer;
}
