// @lynx-js/rspeedy/lib/plugins/output.plugin.js
import { DEFAULT_DIST_PATH_INTERMEDIATE } from '../config/output/dist-path.js';
const defaultFilenameOptions = Object.freeze({
    css: '[name]/[name].css',
});
const defaultDistPathOptions = Object.freeze({
    // We override the default value of Rsbuild(`static/css`) here.
    // Since all the CSS should be encoded into the template in Lynx.
    css: DEFAULT_DIST_PATH_INTERMEDIATE,
});
export function pluginOutput(options) {
    return {
        name: 'lynx:rsbuild:output',
        setup(api) {
            api.modifyRsbuildConfig((config, { mergeRsbuildConfig }) => {
                if (!options) {
                    return mergeRsbuildConfig({
                        output: {
                            filename: {
                                css: defaultFilenameOptions.css,
                            },
                        },
                    }, config);
                }
                return mergeRsbuildConfig(config, {
                    output: {
                        distPath: Object.assign({}, defaultDistPathOptions, options.distPath),
                        filename: Object.assign({}, defaultFilenameOptions, options.filename),
                    },
                });
            });
        },
    };
}
