// @lynx-js/rspeedy/lib/cli/inspect.js
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
import { logger } from '@rsbuild/core';
import { exit } from './exit.js';
import { loadConfig } from '../config/loadConfig.js';
import { createRspeedy } from '../create-rspeedy.js';
export async function inspect(cwd, inspectOptions) {
    try {
        const { content: rspeedyConfig } = await loadConfig({
            cwd,
            configPath: inspectOptions.config,
        });
        const rspeedy = await createRspeedy({ cwd, rspeedyConfig });
        await rspeedy.inspectConfig({
            mode: inspectOptions.mode
                ?? process.env['NODE_ENV']
                ?? 'development',
            verbose: inspectOptions.verbose ?? false,
            outputPath: inspectOptions.output,
            writeToDisk: true,
        });
        return exit();
    }
    catch (error) {
        logger.error(error);
        logger.error();
    }
}
